### PR TITLE
Soundeffects for all the things

### PR DESCRIPTION
**Pull request by Fidaman**

### CODE REVIEW

**Comment by review-ninja** [comment]
<a href="https://app.review.ninja/CUModSquad/cu-patcher-ui/pull/143" target="_blank">ReviewNinja

**Comment by sysrage** [comment]
To get the patch-complete sound to only play after an install/update: https://github.com/sysrage/cu-patcher-ui/tree/soundeffects

File: package.json [hunk 1]
```diff
@@ -49,6 +49,9 @@
     "prebuild:nix": "npm run clean -s && npm run typings",
     "prebuild": "npm run clean -s && npm run typings",
     "publish": "npm run build",
+    "clean:cpui": "rimraf \"%ProgramData%/CSE/Camelot Unchained/cpui/ui\"",
```
Comment (Mehuge) [inline package.json]: This should not be part of this PR.

File: src/ts/PatcherApp.tsx [hunk 2]
```diff
@@ -92,19 +93,22 @@ export class PatcherApp extends React.Component<PatcherAppProps, PatcherState> {
...
     this.props.dispatch(changeRoute(route));
-    this.playSelect();
+    events.fire('play-sound', 'select');
   }
 
   hideChat = () => {
     this.props.dispatch(hideChat());
-  }
+    events.fire('play-sound', 'select');
+}
```
Comment (Mehuge) [inline src/ts/PatcherApp.tsx]: indentation !fix

File: src/ts/PatcherApp.tsx [hunk 3]
```diff
@@ -233,11 +212,7 @@ export class PatcherApp extends React.Component<PatcherAppProps, PatcherState> {
           durationEnter={400} durationLeave={500}>
           {chat}
         </Animate>
-        // Audio
-        <audio src='sounds/select.ogg' ref='sound-select' />
-        <audio src='sounds/launch-game.ogg' ref='sound-launch-game' />
-        <audio src='sounds/patch-complete.ogg' ref='sound-patch-complete' />
-        <audio src='sounds/patcher-theme-v0.1.ogg' ref='sound-bg' />
+        <Sound></Sound>
```
Comment (Mehuge) [inline src/ts/PatcherApp.tsx]: <Sound/>

File: src/ts/Sound.tsx [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+  private playOnce = false;
+
+  getSound(name: string): HTMLAudioElement {
+    return this.refs[name] as HTMLAudioElement;
+  }
+
+  playSound(name: string) {
+    if (!this.props.soundMuted) {
+      let sound = this.getSound(name);
+      sound.play();
```
Comment (Mehuge) [inline src/ts/Sound.tsx]: Should wrap in if (sound) { ... } then if a new sound is added but not yet available it doesn't throw an exception.   !fix

File: src/ts/Sound.tsx [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+      sound.volume = 0.75;
+    }
+  }
+
+  componentDidUpdate() {
+    let soundBg = this.getSound('sound-bg');
+    if (this.props.musicMuted && !soundBg.paused) {
```
Comment (Mehuge) [inline src/ts/Sound.tsx]: Likewise assumes sound-bg is available  !fix

File: src/ts/Sound.tsx [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+      soundBg.pause();
+    } else if (!this.props.musicMuted && soundBg.paused && !this.playOnce) {
+      console.log('playing bg');
+      soundBg.play();
+      soundBg.volume = 0.5;
+    }
+  }
+
+  componentDidMount() {
+    let soundBg = this.getSound('sound-bg');
```
Comment (Mehuge) [inline src/ts/Sound.tsx]: and same here !fix

File: src/ts/Sound.tsx [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+  }
+
+  componentDidUnMount() {
+    events.off('play-sound');
+  }
+
+  render() {
+    return (
+      <div>
+        // Audio
```
Comment (Mehuge) [inline src/ts/Sound.tsx]: Are comments valid here?  Won't this be treated as textContent?

File: src/ts/Sound.tsx [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+      let sound = this.getSound(name);
+      sound.play();
+      sound.volume = 0.75;
+    }
+  }
+
+  componentDidUpdate() {
+    let soundBg = this.getSound('sound-bg');
+    if (this.props.musicMuted && !soundBg.paused) {
+      console.log('pausing bg');
```
Comment (Mehuge) [inline src/ts/Sound.tsx]: loose the console.log !fix

File: src/ts/Sound.tsx [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+      console.log('playing bg');
+      soundBg.play();
+      soundBg.volume = 0.5;
+    }
+  }
+
+  componentDidMount() {
+    let soundBg = this.getSound('sound-bg');
+    if (!this.props.musicMuted) {
+      console.log('playing bg on load');
```
Comment (Mehuge) [inline src/ts/Sound.tsx]: loose the console.log !fix

File: src/ts/sidebar/PatchButton.tsx [hunk 10]
```diff
@@ -147,6 +146,8 @@ class PatchButton extends React.Component<PatchButtonProps, PatchButtonState> {
 
     // Display EULA
     this.setState({ showEuala: true });
+     events.fire('play-sound', 'select');
```
Comment (Mehuge) [inline src/ts/sidebar/PatchButton.tsx]: !fix indent

Comment (Mehuge) [inline src/ts/sidebar/PatchButton.tsx]: !fixed

File: package.json (hunk 1 above)
Comment (Mehuge) [inline package.json]: !fixed

File: src/ts/PatcherApp.tsx (hunk 2 above)
Comment (Mehuge) [inline src/ts/PatcherApp.tsx]: !fixed

File: src/ts/Sound.tsx (hunk 4 above)
Comment (Mehuge) [inline src/ts/Sound.tsx]: !fixed

File: src/ts/Sound.tsx (hunk 5 above)
Comment (Mehuge) [inline src/ts/Sound.tsx]: !fixed

File: src/ts/Sound.tsx (hunk 8 above)
Comment (Mehuge) [inline src/ts/Sound.tsx]: !fixed

File: src/ts/Sound.tsx (hunk 9 above)
Comment (Mehuge) [inline src/ts/Sound.tsx]: !fixed
